### DISCUSSION TITLE
Improvements to Stacks customizations

### DISCUSSION THREAD

**Discussion by nickytonline** [remix-run/remix#5544] [Proposals]
I've been working on the [remix.init](https://remix.run/docs/en/v1/pages/stacks#customize-initialization) script for the Netlify Remix work (see https://github.com/netlify/remix-template/pull/83/files#diff-47d697bbfe7441205f5569284242c952cab65c0d9564344a81c9f10d42673ec3)

While working on this, I did some things that I think would make sense to be baked into Stacks customization:
- the ability to exclude folders from the template repository. For example, in our case, our `.github` folder contains workflows we use for CI/CD, but having it in the generated template makes no sense. Currently have an array of exclude folders and remove those folders from the generated template,

This could be defined in a config or, as my coworker @ascorbic suggested, an ignore file, like .gitignore but for Stacks.

- Not a showstopper by any means, but installing inquirer for customizations seems unnecessary. It could be passed in via the `main` function

`remit.init/index.js`

```diff
- async function main({ rootDirectory, isTypeScript }) {
+ async function main({ rootDirectory, isTypeScript, inquirer }) {
  // customization code
  // ...
}
```

- Run conversion to JavaScript after the customizations via remix/init/index.js have been completed. We have files in remix.init that get converted to *.js and *.jsx when a user selects JavaScript from the Remix CLI. Currently, I handle this by manipulating file extensions in my remix.init folder that gets copied. e.g.

```js
const tsExtensionMatcher = /\.ts(x?)$/;

function convertToJsExtension(file) {
  return file.replace(tsExtensionMatcher, ".js$1");
}

async function copyTemplateFiles({ files, rootDirectory, isTypeScript }) {
  for (const [file, target] of files) {
    let sourceFile = file;
    let targetFile = target || file;

    // change the target file extension .tsx to .jsx only if the project has been converted to JavaScript
    if (!isTypeScript && file.match(tsExtensionMatcher)) {
      // If they chose JavaScript, the source file is converted to .js or .jsx and
      // we need the target file to be .js or .jsx for the same reason.
      sourceFile = convertToJsExtension(file);
      targetFile = convertToJsExtension(targetFile);
    }

    await fs.copyFile(
      join(rootDirectory, "remix.init", sourceFile),
      join(rootDirectory, targetFile)
    );
  }
}
```

https://github.com/netlify/remix-template/blob/e7122a1bad1030d06535626c0e5abccfadb53c7f/remix.init/index.js#L22-L46

**Comment by mcansh** [comment]
took a first quick stab at a `.remixignore` file here, but for all `npx create-remix` use cases https://github.com/remix-run/remix/pull/5551

**Comment by nickytonline** [comment]
One other thing I noticed is if you do not say yes to installing npm dependencies, the remix.init script never runs. Should I file an issue for that?

**Reply by mcansh** [reply]
yeah being that you may need dependencies in your `remix.init`, we won't run it at all, but you can use `remix init` to manually run it. looks like that command is missing from the docs 😅

**Reply by mcansh** [reply]
it's in the CLI help menu though - i'll get a PR sorted to bring it over

```log
  Initialize a project::

    Remix project templates may contain a `remix.init` directory
    with a script that initializes the project. This script automatically
    runs during `remix create`, but if you ever need to run it manually
    (e.g. to test it out) you can:

    $ remix init
```

**Reply by nickytonline** [reply]
Ahh gotcha, so if they said no to installing npm dependencies, once the template generates the project they just run `npx remix init` from the root of the new project?

**Reply by mcansh** [reply]
😇 https://github.com/remix-run/remix/pull/5651

**Reply by mcansh** [reply]
> Ahh gotcha, so if they said no to installing npm dependencies, once the template generates the project they just run `npx remix init` from the root of the new project?

yup exactly
